### PR TITLE
Simple sampling (the same as custody columns) applied to forkchoice on block import

### LINKED ISSUE #104: Update forkchoice and sampling according to the spec
We should assert data available to import the block https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip7594/fork-choice.md#modified-is_data_available
But sampling is equal to the custody (almost) https://github.com/ethereum/consensus-specs/blob/dev/specs/_features/eip7594/das-core.md#subnet-sampling

### PR DESCRIPTION
**Pull request by zilm13**
Fixes #104

### CODE REVIEW

**Comment by zilm13** [comment]
there are issues with Custody which should be fixed before merging this

**Comment by zilm13** [comment]
custody issues are fixed

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/statetransition/availability/DataAndValidationResult.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+package tech.pegasys.teku.spec.logic.common.statetransition.availability;
+
+import com.google.common.base.MoreObjects;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+
+public class DataAndValidationResult<Data> {
```
Comment (Nashatyrev) [inline ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/statetransition/availability/DataAndValidationResult.java]: Could it be a `record` ?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImpl.java [hunk 2]
```diff
@@ -103,6 +108,9 @@ public DataColumnSidecarCustodyImpl(
     this.nodeId = nodeId;
     this.totalCustodySubnetCount = totalCustodySubnetCount;
     this.eip7594StartEpoch = spec.getForkSchedule().getFork(SpecMilestone.EIP7594).getEpoch();
+    this.knownSavedIdentifiers =
```
Comment (Nashatyrev) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImpl.java:111]: May be would better move this optimization down to either  
- `DataColumnSidecarDbImpl`
- `CombinedChainDataClient`
- Or make a dedicated `DataColumnSidecarDb` wrapper to cache this?

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasic.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+  @Override
+  public void onSlot(UInt64 slot) {}
+
+  @Override
+  public void onNewFinalizedCheckpoint(Checkpoint checkpoint, boolean fromOptimisticBlock) {
+    db.getFirstSamplerIncompleteSlot()
+        .thenCompose(
+            maybeSlot ->
+                maybeSlot.map(db::setFirstSamplerIncompleteSlot).orElse(SafeFuture.COMPLETE))
```
Comment (Nashatyrev) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasic.java:105]: Not sure I understand what we are doing here...

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/DataColumnSidecarAvailabilityChecker.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      validationResult.complete(DataAndValidationResult.notRequired());
+      LOG.info("Availability check for slot {} NOT_REQUIRED, epoch too old ", block.getSlot());
+      return true;
+    }
+
+    dataAvailabilitySampler
+        .checkDataAvailability(block.getSlot(), block.getRoot(), block.getParentRoot())
+        .finish(
+            dataColumnSidecars ->
+                validationResult.complete(validateImmediately(dataColumnSidecars)),
```
Comment (Nashatyrev) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/DataColumnSidecarAvailabilityChecker.java:94]: Should we really validate sadecars here? I suppose that any sidecar returned by the custody should be implicitly valid

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasic.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                miscHelpersEip7594.computeCustodyColumnIndexes(nodeId, totalCustodySubnetCount));
+    if (columnIndexes.isEmpty()) {
+      return SafeFuture.completedFuture(List.of());
+    }
+
+    return SafeFuture.collectAll(
+            columnIndexes.get().stream()
+                .map(
+                    index ->
+                        custody.getCustodyDataColumnSidecar(
```
Comment (Nashatyrev) [inline ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasic.java:79]: Assuming that all columns in our DB are inherently valid, may be it is cheaper to just query for available column indexes?

**Review by Nashatyrev — APPROVED** [review]
LGTM
2 issues to be created to address comments

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImpl.java (hunk 2 above)
Comment (zilm13) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImpl.java:111]: https://github.com/Nashatyrev/teku/issues/122

File: ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasic.java (hunk 5 above)
Comment (zilm13) [reply ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasSamplerBasic.java:79]: https://github.com/Nashatyrev/teku/issues/123

File: ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/statetransition/availability/DataAndValidationResult.java (hunk 1 above)
Comment (zilm13) [reply ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/common/statetransition/availability/DataAndValidationResult.java]: Updated

**Pull request merged**